### ISSUE TITLE
Hystrix Circuit Never Closes After Opening

### ISSUE THREAD

**Issue by ryanjbaxter** [spring-cloud/spring-cloud-netflix#1204]
I am seeing that circuit breakers that open do to a failure in a downstream service remain open and never close again when the downstream service is once again functioning.  I am able to reproduce this in Brixton.RELEASE but not in the latest Angel release.

To reproduce the problem clone these three git projects and run them locally by running `mvn spring-boot:run`.
https://github.com/ryanjbaxter/ocr-eureka
https://github.com/ryanjbaxter/ocr-races
https://github.com/ryanjbaxter/ocr-participants

If you hit `http://localhost:8282/participant`s that will call the `/participants` route in the `ocr-races` service.  Notice this will end up calling a [method protected by a circuit breaker](https://github.com/ryanjbaxter/ocr-races/blob/master/src/main/java/com/ryanjbaxter/spring/cloud/ocr/races/OcrRacesApplication.java#L74).  If you then shut down the `ocr-participants` service and continue hitting `/participants` the circuit will open as expected.  If you now restore the `ocr-participants` service and continue hitting `/participants` the circuit will never close again, no matter how long you wait.

Further experimentation has revealed that this only occurs when using a Feign client.  If you use a RestTemplate (load balanced or just a vanilla RestTemplate) to call the downstream service than the circuit opens and closes as expected.

**Comment by dsyer** [comment]
The sample here doesn't use Feign: https://github.com/spring-cloud-samples/tests/tree/master/hystrix and it has the same problem

**Comment by ryanjbaxter** [comment]
After looking at the sample from @dsyer we determined that is was not actually ever going to close the circuit because there was no wail that the `fail` endpoint would every produce a successful response.  We added the ability to pass a header causing the endpoint to produce a successful response.  After doing that we determined that the sample does not reproduce the problem.

**Comment by ryanjbaxter** [comment]
After deeper investigation the problem with the sample appears to be due to new functionality added in Brixton.  In Brixton methods in Feign Clients are now automatically wrapped in Hystrix Commands when Hystrix is present on the classpath.  See http://cloud.spring.io/spring-cloud-static/spring-cloud.html#spring-cloud-feign-hystrix.

The reason why the circuit never closes in the samples above is because the `@FeignClient` annotation does not contain a fallback parameter.  When the `participants` service is no longer available Hystrix would try to call the fallback method, but since no fallback was defined it would throw an Exception.  This Exception is thrown in `HystrixCommand.getFallback`.  In order for the circuit to close Hystrix needs a certain number of requests to succeed as well as a certain amount of time to elapse.  However we never get to the point of even trying to let any request through the circuit because the Exception was being thrown due to the missing fallback, therefore the circuit will never close.  Once a fallback is defined in the FeignClient the circuit will open and close as expected.

The code in the sample really should have further changes though.  In Brixton we should completely remove the defaultFallback parameter from the HystrixCommand, in fact the entire `ParticipantsBean` can be eliminated and the `/participants` endpoint can use the FeignClient directly since it is now protected by Hystrix Commands.

The problem here of course is more of a migration issue from Angel -> Brixton.  Applications following this pattern (using FeignClients wrapped in Hystrix Commands) will experience the same behavior.  You can disable Hystrix support in FeignClients by setting `feign.hystrix.enabled` to false in Brixton and that will keep the same behavior/functionality present in Angel.

**Comment by adriancole** [comment]
Here's an example exception complements of @ryanjbaxter!

```bash
java.lang.UnsupportedOperationException: No fallback available.
    at com.netflix.hystrix.HystrixCommand.getFallback(HystrixCommand.java:284) ~[hystrix-core-1.5.3.jar:1.5.3]
    at feign.hystrix.HystrixInvocationHandler$1.getFallback(HystrixInvocationHandler.java:108) ~[feign-hystrix-8.16.2.jar:8.16.2]
    at com.netflix.hystrix.HystrixCommand$2.call(HystrixCommand.java:307) ~[hystrix-core-1.5.3.jar:1.5.3]
    at com.netflix.hystrix.HystrixCommand$2.call(HystrixCommand.java:303) ~[hystrix-core-1.5.3.jar:1.5.3]
    at rx.internal.operators.OnSubscribeDefer.call(OnSubscribeDefer.java:46) [rxjava-1.1.5.jar:1.1.5]
    at rx.internal.operators.OnSubscribeDefer.call(OnSubscribeDefer.java:35) [rxjava-1.1.5.jar:1.1.5]
    at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:50) [rxjava-1.1.5.jar:1.1.5]
    at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30) [rxjava-1.1.5.jar:1.1.5]
    at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:50) [rxjava-1.1.5.jar:1.1.5]
    at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30) [rxjava-1.1.5.jar:1.1.5]
    at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:50) [rxjava-1.1.5.jar:1.1.5]
    at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30) [rxjava-1.1.5.jar:1.1.5]
    at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:50) [rxjava-1.1.5.jar:1.1.5]
    at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30) [rxjava-1.1.5.jar:1.1.5]
    at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:50) [rxjava-1.1.5.jar:1.1.5]
    at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30) [rxjava-1.1.5.jar:1.1.5]
    at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:50) [rxjava-1.1.5.jar:1.1.5]
    at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30) [rxjava-1.1.5.jar:1.1.5]
```

**Comment by adriancole** [comment]
so we need something (probably feign) to set fallbackEnabled=false when we know there's no fallback type. Otherwise, `HystrixCommand.toObservable()` assumes a fallback is present. Since it isn't there's a very late `UnsupportedOperationException` which is the thing that's mucking things up.

**Comment by adriancole** [comment]
So thinking about this.. `HystrixCommand.getFallback()` is meant to be implemented, not called directly, right? the UnsupportedOperationException is part of the fallback contract.. ex there's specific logic in `HystrixCommand (named handleFallbackError)` that throws `HystrixRuntimeException` on unsupported.

If ...
* no-one is calling `HystrixCommand.getFallback()` except Hystrix
* Hystrix expects `HystrixCommand.getFallback()` to throw `UnsupportedOperationException`
Why are we seeing a stacktrace and suspecting this as the cause of an open circuit.

I wonder if the stack trace were longer, if we could identify what's up.. ex by adding `--logging.exception-conversion-word=%wEx{full}` to dump full stack traces instead of truncated ones.

**Comment by adriancole** [comment]
I think the `UnsupportedOperationException` from `HystrixCommand.getFallback` is distracting, as it is the default behavior. Perhaps there's another explanation from this or some way to make a straight-forward test case that shows the circuit left-open problem without starting several processes.

**Comment by ryanjbaxter** [comment]
It occurred to me over the weekend that in my example I was seeing something very strange when looking at the Hystrix Dashboard.  Even though there was 2 circuit breakers in my races service, there was only 1 circuit breaker showing up in the dashboard, it was called `getParticipants`.
![image](https://cloud.githubusercontent.com/assets/524254/17113316/2a220628-5278-11e6-8574-2189a2c900a2.png)
 I then came to the realization that my `@HystrixCommand` and my FeignClient both had the same method name and signature.  This made me think that maybe Hystrix was therefore confusing the fact that there were two commands because both commands would be using the same key, in this case `getParticipants`.

I dug deeper into Hystrix and found this to be the case.  By default the command key is the method name.

> By default the name of command key is command method name: getUserById, default group key name is class name of annotated method: UserService. You can change it using necessary @HystrixCommand properties:

Hystrix will cache commands as we can see here
https://github.com/Netflix/Hystrix/blob/master/hystrix-core/src/main/java/com/netflix/hystrix/HystrixCircuitBreaker.java#L77
So in the case where you have to commands with the same method name, the cache will return the  same circuit breaker for both commands.

In my example this was always the circuit for the `@HystrixCommand` and not the one for the FeignClient.  When the circuit would open up due to an error in the downstream service, both circuit breakers for both commands would be open.  However when the downstream service was restored and we should see the circuit did not close.  Now however the reason it was staying open, in my example anyways, was because my FeignClient does not have a fallback which was also throwing an exception again marking the circuit as being open.  

Now if I change the name of the methods either in the `@HystrixCommand` or in the interface for the FeignClient than I end up with two distinct circuits.  This allows for the correct circuit breaker to be marked as open for each hystrix command and you no longer see the problem of the circuit never closing because the circuit for the `FeignClient` never is marked as open so the fallback logic is never attempted to be executed, hence the exception due to the missing fallback is never thrown.

**Comment by ryanjbaxter** [comment]
I created an issue in Hystrix around the circuit breaker keys not being unique enough https://github.com/Netflix/Hystrix/issues/1293

**Comment by ryanjbaxter** [comment]
This should be fixed by #1273.

This contains the fix from Feign https://github.com/OpenFeign/feign/issues/434

**Comment by ryanjbaxter** [comment]
Fixed via #1278
